File: +ndi/+database/+fun/openMINDSobj2ndi_document.m
function d = openMINDSobj2ndi_document(openmindsObj, session_id, dependency_type, dependency_value, varargin)
% OPENMINDSOBJ2NDI_DOCUMENT - openMinds objects to set of ndi.document objects
%
% D = ndi.database.fun.openMINDSobj2ndi_document(OPENMINDS_OBJ, SESSION_ID, [DEPENDENCY_TYPE], [DEPENDENCY_VALUE])
%
% Convert a cell array of openMINDS objects to a set of ndi.document objects.
% D is a cell array of ndi.document objects. If the document is requested to be of a particular
% DEPENDENCY_TYPE ('subject', 'element'), then the corresponding dependency is set to
% DEPENDENCY_VALUE.
%
% Example 1:
%   p = personWithTwoAffiliations(); % openMINDS library function
%   session_id = S.id(); % get the id of an ndi.session S
%   d = ndi.database.fun.openMINDSobj2ndi_document(p, session_id);
%
% Example 2:
%   s = openminds.controlledterms.Species('name','Mustela putorius furo','preferredOntologyIdentifier','NCBI:txid9669');
%   session_id = S.id(); % get the id of an ndi.session S
%   subject_docs = S.database_search(ndi.query('','isa','subject'));
%   d = ndi.database.fun.openMINDSobj2ndi_document(s, session_id, 'subject', subject_docs{1}.id());
% 


if ~iscell(openmindsObj),
    newcell = {};
    for i=1:numel(openmindsObj)
        newcell{i} = openmindsObj(i);
    end;
    openmindsObj = newcell;
end;

s = ndi.database.fun.openMINDSobj2struct(openmindsObj);

if nargin<3,
	dependency_type = '';
end;

if nargin<4,
	dependency_value = '';
end;

if ~isempty(dependency_type) & isempty(dependency_value),
	error(['DEPENDENCY_VALUE must not be empty if DEPENDENCY_TYPE is given.']);
end;

docName = 'openminds';

switch lower(dependency_type),
	case '',
		dependency_name = '';
		docName = 'openminds';
	case 'subject',
		dependency_name = 'subject_id';
		docName = 'openminds_subject';
	case 'element',
		dependency_name = 'element_id';
		docName = ['openminds_element'];
	case 'stimulus',
		dependency_name = 'stimulus_element_id';
		docName = ['openminds_stimulus'];
	otherwise,
		error(['Unknown DEPENDENCY_TYPE ' dependency_type '.']);
end;

d = {};

for i=1:numel(s),
	openminds_struct = rmfield(s(i),'complete');
	ndi_id_here = openminds_struct.ndi_id;
	openminds_struct = rmfield(openminds_struct,'ndi_id');
	d{i} = ndi.document(docName,'base.id',ndi_id_here,...
		'base.session_id',session_id,...
		'openminds',openminds_struct,varargin{:});
	fn = fieldnames(openminds_struct.fields);
	added_dependency = 0;
	for j=1:numel(fn),
		g = getfield(openminds_struct.fields,fn{j});
		if iscell(g),
			for k=1:numel(g),
				if ischar(g{k}),
					if startsWith(g{k},'ndi://'),
						id_here = g{k}(7:end);
						d{i} = add_dependency_value_n(d{i},...
							'openminds',id_here,...
							'ErrorIfNotFound',0);
						added_dependency = 1;
					end;
				end;
			end;
		end;
	end;
	if ~added_dependency,
		d{i} = set_dependency_value(d{i},'openminds','','ErrorIfNotFound',0);
	end;
	if ~isempty(dependency_name),
		d{i} = d{i}.set_dependency_value(dependency_name,dependency_value);
	end;
end;

File: ndi-python/ndi/db/fun/openMINDSobj2ndi_document.py
"""
Convert openMINDS objects to NDI document objects.

This module provides functionality to convert openMINDS metadata objects to
ndi.document objects for storage in the NDI database.

Ported from MATLAB: src/ndi/+ndi/+database/+fun/openMINDSobj2ndi_document.m
"""

from typing import List, Optional, Union


def openMINDSobj2ndi_document(
    openminds_obj,
    session_id: str,
    dependency_type: str = '',
    dependency_value: str = ''
) -> List:
    """
    Convert openMINDS objects to a set of ndi.document objects.

    Args:
        openminds_obj: A single openMINDS object or cell array/list of openMINDS objects
        session_id: The session ID to associate with the documents
        dependency_type: Optional dependency type ('subject', 'element', 'stimulus')
        dependency_value: Value for the dependency (required if dependency_type is given)

    Returns:
        List of ndi.document objects

    Raises:
        ValueError: If dependency_type is given but dependency_value is empty
        ImportError: If openMINDS library is not available

    Example:
        >>> # Assuming openminds_obj is an openMINDS object
        >>> session_id = session.id()
        >>> docs = openMINDSobj2ndi_document(openminds_obj, session_id)
        >>> # With dependency
        >>> subject_id = subject_doc.id()
        >>> docs = openMINDSobj2ndi_document(species_obj, session_id, 'subject', subject_id)

    Notes:
        - Requires openMINDS Python library (https://github.com/openMetadataInitiative/openMINDS_Python)
        - Creates documents with openminds metadata structure
        - Automatically detects and links dependencies between openMINDS objects
        - Dependency types map to NDI document dependency names:
          - 'subject' -> 'subject_id'
          - 'element' -> 'element_id'
          - 'stimulus' -> 'stimulus_element_id'
    """
    from ndi.document import Document
    from .openMINDSobj2struct import openMINDSobj2struct

    # Validate dependency arguments
    if dependency_type and not dependency_value:
        raise ValueError("DEPENDENCY_VALUE must not be empty if DEPENDENCY_TYPE is given")

    # Ensure openminds_obj is a list
    if not isinstance(openminds_obj, list):
        if hasattr(openminds_obj, '__iter__') and not isinstance(openminds_obj, str):
            openminds_obj = list(openminds_obj)
        else:
            openminds_obj = [openminds_obj]

    # Convert openMINDS objects to structures
    structures = openMINDSobj2struct(openminds_obj)

    # Determine document name and dependency name based on type
    dependency_name = ''
    doc_name = 'openminds'

    if dependency_type:
        dependency_type_lower = dependency_type.lower()
        if dependency_type_lower == 'subject':
            dependency_name = 'subject_id'
            doc_name = 'openminds_subject'
        elif dependency_type_lower == 'element':
            dependency_name = 'element_id'
            doc_name = 'openminds_element'
        elif dependency_type_lower == 'stimulus':
            dependency_name = 'stimulus_element_id'
            doc_name = 'openminds_stimulus'
        else:
            raise ValueError(f"Unknown DEPENDENCY_TYPE: {dependency_type}")

    # Create NDI documents from structures
    docs = []
    for struct in structures:
        # Remove 'complete' field if present
        openminds_struct = {k: v for k, v in struct.items() if k != 'complete'}

        # Extract NDI ID
        ndi_id = openminds_struct.pop('ndi_id', None)

        # Create document
        doc = Document(
            doc_name,
            **{
                'base.id': ndi_id,
                'base.session_id': session_id,
                'openminds': openminds_struct
            }
        )

        # Add dependencies for nested openMINDS objects
        added_dependency = False
        fields = openminds_struct.get('fields', {})
        if isinstance(fields, dict):
            for field_name, field_value in fields.items():
                # Handle cell arrays/lists
                if isinstance(field_value, list):
                    for item in field_value:
                        if isinstance(item, str) and item.startswith('ndi://'):
                            # Extract ID from NDI URI
                            ref_id = item[6:]  # Remove 'ndi://' prefix
                            doc = doc.add_dependency_value_n('openminds', ref_id)
                            added_dependency = True

        # If no dependencies were added, set empty openminds dependency
        if not added_dependency:
            try:
                doc = doc.set_dependency_value('openminds', '')
            except:
                pass  # Ignore if method doesn't exist

        # Add the specified dependency if provided
        if dependency_name and dependency_value:
            doc = doc.set_dependency_value(dependency_name, dependency_value)

        docs.append(doc)

    return docs
